grpc/private/gateway.go

```diff
@@ -0,0 +1,83 @@
+package private
+
+import (
+	"bytes"
+	"net"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/keratin/authn-server/app"
+	authnpb "github.com/keratin/authn-server/grpc"
+	"github.com/keratin/authn-server/grpc/internal/gateway"
+	"github.com/keratin/authn-server/grpc/public"
+	"github.com/keratin/authn-server/server/views"
+	"golang.org/x/net/context"
+	grpc "google.golang.org/grpc"
+)
+
+func RunPrivateGateway(ctx context.Context, app *app.App, r *mux.Router, conn *grpc.ClientConn, l net.Listener) error {
+
+	gmux := runtime.NewServeMux(
+		runtime.WithForwardResponseOption(gateway.CookieSetter(app.Config)), // Cookies always have to go first
```

Comment:
Yes

Reply:
How do you feel about that setup? I backed my way into it because I wanted to maintain backwards compatibility during 1.x while still supporting a reduced access public port. I'm currently leaning towards simplifying in 2.x.

Meanwhile, the gRPC server doesn't have any history to maintain so it can make its own choice here.

Reply:
I think it's fine. To me it means I can have my public users and back-office users both exist on the same dataset, and what they're able to do is ruled by another authz microservice. It will also allow Keratin AuthN to be reused for internal apps (without having to go through the public IP address) without having to run separate instances.

Reply:
But do you think that the private port should bother to continue handling public endpoints? Currently the split setup looks like:

* public port -- handles public endpoints
* private port -- handles private _and public_ endpoints

The only reason the private port works that way currently is because I didn't want to introduce a breaking change in 1.x for single-port deployments. In 2.x I may switch to enforcing the two port setup.

Meanwhile, the gRPC server isn't burdened by existing deployments. It could avoid mixing public routes into the private gateway from the beginning and require a two-port configuration. Do you think there's an argument for waiting until 2.x?

Reply:
Found my own answer: backwards compatibility remains valuable to the gRPC server so that it can be a drop-in replacement (albeit experimental). Happy to defer simplifying all of this until 2.x.